--- Day 2: Rock Paper Scissors ---

The Elves begin to set up camp on the beach. To decide whose tent gets to be closest to the snack storage, a giant Rock Paper Scissors tournament is already in progress.

Rock Paper Scissors is a game between two players. Each game contains many rounds; in each round, the players each simultaneously choose one of Rock, Paper, or Scissors using a hand shape. Then, a winner for that round is selected: Rock defeats Scissors, Scissors defeats Paper, and Paper defeats Rock. If both players choose the same shape, the round instead ends in a draw.

Appreciative of your help yesterday, one Elf gives you an encrypted strategy guide (your puzzle input) that they say will be sure to help you win. "The first column is what your opponent is going to play: A for Rock, B for Paper, and C for Scissors. The second column--" Suddenly, the Elf is called away to help with someone's tent.

The second column, you reason, must be what you should play in response: X for Rock, Y for Paper, and Z for Scissors. Winning every time would be suspicious, so the responses must have been carefully chosen.

The winner of the whole tournament is the player with the highest score. Your total score is the sum of your scores for each round. The score for a single round is the score for the shape you selected (1 for Rock, 2 for Paper, and 3 for Scissors) plus the score for the outcome of the round (0 if you lost, 3 if the round was a draw, and 6 if you won).

Since you can't be sure if the Elf is trying to help you or trick you, you should calculate the score you would get if you were to follow the strategy guide.

For example, suppose you were given the following strategy guide:

    A Y
    B X
    C Z

This strategy guide predicts and recommends the following:

In the first round, your opponent will choose Rock (A), and you should choose Paper (Y). This ends in a win for you with a score of 8 (2 because you chose Paper + 6 because you won).
In the second round, your opponent will choose Paper (B), and you should choose Rock (X). This ends in a loss for you with a score of 1 (1 + 0).
The third round is a draw with both players choosing Scissors, giving you a score of 3 + 3 = 6.
In this example, if you were to follow the strategy guide, you would get a total score of 15 (8 + 1 + 6).

What would your total score be if everything goes exactly according to your strategy guide?

In [15]:
strategy = list()
with open("strategy.txt", "r") as f:
    for line in f:
        strategy.append(line.split())
strategy[:3]

[['B', 'X'], ['B', 'Y'], ['A', 'Y']]

In [73]:
# A and X = Rock, B and Y = Paper, C and Z = Scissors
outcome_scores = {
    "AX": 3,  # draw
    "AY": 6,  # win
    "AZ": 0,  # lose
    "BX": 0,
    "BY": 3,
    "BZ": 6,
    "CX": 6,
    "CY": 0,
    "CZ": 3,
}


def round_score(pair):
    """Returns the score of a round.
        A = Rock
        B = Paper
        C = Scissors

        X = Rock
        Y = Paper
        Z = Scissors
    """

    my_choice = ord(pair[1]) - 87
    round_str = "".join(pair)
    my_score = outcome_scores[round_str]
    my_total_score = my_choice + my_score

    return my_total_score

round_score(['C', 'Z'])

6

In [74]:
my_total_score = 0
their_total_score = 0
for round in strategy:
    my_score = round_score(round)
    my_total_score += my_score

print(my_total_score)

15691


Attempts:

I thought it was not certain that XYZ were rock, paper and scissors. So I made a rotation and then chose the outcome that would give me the lowest winning margin to make it less obvious. That means I also calculated my opponent's score...

But I was wrong. XYZ were indeed rock, paper and scissors. I fixed the code accordingly to only get 1 response; the correct one.

- 12489
- 13522
- 15691 --> correct

--- Part Two ---

The Elf finishes helping with the tent and sneaks back over to you. "Anyway, the second column says how the round needs to end: X means you need to lose, Y means you need to end the round in a draw, and Z means you need to win. Good luck!"

The total score is still calculated in the same way, but now you need to figure out what shape to choose so the round ends as indicated. The example above now goes like this:

In the first round, your opponent will choose Rock (A), and you need the round to end in a draw (Y), so you also choose Rock. This gives you a score of 1 + 3 = 4.
In the second round, your opponent will choose Paper (B), and you choose Rock so you lose (X) with a score of 1 + 0 = 1.
In the third round, you will defeat your opponent's Scissors with Rock for a score of 1 + 6 = 7.
Now that you're correctly decrypting the ultra top secret strategy guide, you would get a total score of 12.

Following the Elf's instructions for the second column, what would your total score be if everything goes exactly according to your strategy guide?

In [62]:
winning_move = {
    "A": "B",   # if I see rock I play paper
    "B": "C",
    "C": "A",
}

losing_move = {
    "A": "C",   # if I see rock I play scissors
    "B": "A",
    "C": "B",
}

In [75]:
def decide_my_move(opponent, outcome):
    """Returns the move I should make based on the opponent's move and the outcome of the round.
        X = lose
        Y = draw
        Z = win
    """

    if outcome == "X":
        return chr(ord(losing_move[opponent]) + 23)   # e.g. convert "A" to "Z"

    if outcome == "Y":
        return chr(ord(opponent) + 23)   # e.g. convert "A" to "X"
    
    if outcome == "Z":
        return chr(ord(winning_move[opponent]) + 23)   # e.g. convert "A" to "Y"

decide_my_move(*["C", "Z"])

'X'

In [79]:
def round_score(pair):
    """Returns the score of a round.
    """

    my_choice = decide_my_move(*pair)
    my_choice_score = ord(my_choice) - 87
    round_str = "".join(pair[0] + my_choice)
    my_score = outcome_scores[round_str]
    my_total_score = my_choice_score + my_score

    return my_total_score

round_score(['B', 'X'])

1

In [80]:
my_total_score = 0
their_total_score = 0
for round in strategy:
    my_score = round_score(round)
    my_total_score += my_score

print(my_total_score)

12989
